acir/src/native_types/witness.rs

```diff
@@ -27,23 +27,75 @@ impl Witness {
     pub const fn can_defer_constraint(&self) -> bool {
         true
     }
+}
+
+impl From<u32> for Witness {
+    fn from(value: u32) -> Self {
+        Self(value)
+    }
+}
+
+/// A map from the witnesses in a constraint system to the field element values
+#[derive(Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord, Default, Serialize, Deserialize)]
+pub struct WitnessMap(BTreeMap<Witness, acir_field::FieldElement>);
+
+impl WitnessMap {
+    pub fn new() -> Self {
+        Self(BTreeMap::new())
+    }
+    pub fn get(&self, witness: &Witness) -> Option<&acir_field::FieldElement> {
+        self.0.get(witness)
+    }
+    pub fn get_index(&self, index: u32) -> Option<&acir_field::FieldElement> {
+        self.0.get(&index.into())
+    }
+    pub fn contains_key(&self, key: &Witness) -> bool {
+        self.0.contains_key(key)
+    }
+    pub fn insert(
+        &mut self,
+        key: Witness,
+        value: acir_field::FieldElement,
+    ) -> Option<acir_field::FieldElement> {
+        self.0.insert(key, value)
+    }
+}
 
-    pub fn to_bytes(
-        witnesses: &std::collections::BTreeMap<Witness, acir_field::FieldElement>,
-    ) -> Vec<u8> {
-        let buf = rmp_serde::to_vec(witnesses).unwrap();
+impl Index<&Witness> for WitnessMap {
+    type Output = acir_field::FieldElement;
+
+    fn index(&self, index: &Witness) -> &Self::Output {
+        &self.0[index]
+    }
+}
+
+impl From<BTreeMap<Witness, acir_field::FieldElement>> for WitnessMap {
+    fn from(value: BTreeMap<Witness, acir_field::FieldElement>) -> Self {
+        Self(value)
+    }
+}
+
+impl From<WitnessMap> for Vec<u8> {
```

Comment:
I'm also unsure whether we want to keep this serialisation method in future as we've removed all the other instances of it: https://github.com/noir-lang/acvm/issues/186

Reply:
Will that mean that `read` and `write` need to be implemented on the WitnessMap type? I believe that is what is used for Circuits now

Reply:
Nah, I think at most we'd keep this as an ad hoc method so we can add a deprecation notice in future.

Reply:
Hm 🤔  I'm currently thinking that the backend boundary needs to be bytes (so, `Vec<u8>`/`&[u8]`/`Bytes`) and nargo serializes it and a backend deserializes it using the ACVM crate, so I was pretty sure we'd need `read` and `write` functions

Reply:
I don't follow. This serialisation method is only used when saving the witness to disk rather than when communicating with backends.

Reply:
Let's keep these for now since removing them (and the follow up discussion) is unrelated to this PR